torch/_inductor/codegen/triton.py

```diff
@@ -2395,63 +2399,124 @@ def scan(
-            reduced_size = f"[{', '.join(reduced_size)}]"
+            if self.persistent_reduction:
+                masked_value = where_cond(value)
+                masked_values.append(masked_value)
+            else:
+                accumulator = self.cse.newvar()
+                reduced_size = self.dense_size_list()
+                reduced_size[-1] = "1"
+                reduced_size = f"[{', '.join(reduced_size)}]"
 
-            self.body.writeline(
-                f"{accumulator} = tl.full({reduced_size}, {default}, {acc_type})"
-            )
+                self.body.writeline(
+                    f"{accumulator} = tl.full({reduced_size}, {default}, {acc_type})"
+                )
 
-            masked_value = where_cond(value)
-            partial_reduce = self.cse.generate(
-                self.compute,
-                self.reduction_resize(
-                    f"tl.reduce({value}, {dim}, {combine_helper_fn})"
-                ),
-            )
-            acc_next = combine_fn(accumulator, partial_reduce)
-            partial_scan = self.cse.generate(
-                self.compute,
-                f"tl.associative_scan({masked_value}, {dim}, {combine_helper_fn})",
-            )
-            result_var = self.cse.generate(
-                self.compute, combine_fn(accumulator, partial_scan)
+                masked_value = where_cond(value)
+
+                masked_values.append(masked_value)
+                accumulators.append(accumulator)
+
+        def csv(values):
+            return " ".join(f"{value}," for value in values)
+
+        def cse_multiple(line, n, masks):
+            cache_keys = [f"{line}, {i}, {masks}" for i in range(n)]
+            if all(cache_key in self.cse.cache for cache_key in cache_keys):
+                return [self.cse.cache[cache_key] for cache_key in cache_keys]
+            result_vars = [self.cse.newvar() for _ in range(n)]
+            self.compute.writeline(
+                f"{csv(result_vars)} = {line}",
             )
-            self.compute.writeline(f"{accumulator} = {acc_next}")
+            for result_var, cache_key in zip(result_vars, cache_keys):
+                if masks:
+                    result_var.mask_vars = masks  # type: ignore[attr-defined]
+                self.cse.cache[cache_key] = result_var
+            return tuple(result_vars)
+
+        result_vars = cse_multiple(
+            f"tl.associative_scan(({csv(masked_values)}), {dim}, {combine_helper_fn})",
+            len(values),
+            masks,
+        )
 
-        result_var.mask_vars = masks  # type: ignore[attr-defined]
-        return result_var
+        if not self.persistent_reduction:
+            # TODO(isuruf): remove the else branch when rocm-triton pin is updated
+            if torch.version.hip is None:
+                partial_reduce = cse_multiple(
+                    f"tl.reduce(({csv(broadcasted_values)}), {dim}, {combine_helper_fn}, keep_dims=True)",
```

Comment:
Is it just the `keep_dims` part that's not compatible with rocm? If so, you can just call `self.reduction_resize(f"tl.reduce...` as it did before and it will work for both rocm and cuda.